optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -867,6 +867,146 @@
+     * @param groupKeyBMapping never null, function to convert the original tuple into a second fact
+     * @param groupKeyCMapping never null, function to convert the original tuple into a third fact
+     * @param groupKeyDMapping never null, function to convert the original tuple into a fourth fact
+     * @param <GroupKeyA_> the type of the first fact in the destination {@link QuadConstraintStream}'s tuple
+     * @param <GroupKeyB_> the type of the second fact in the destination {@link QuadConstraintStream}'s tuple
+     * @param <GroupKeyC_> the type of the third fact in the destination {@link QuadConstraintStream}'s tuple
+     * @param <GroupKeyD_> the type of the fourth fact in the destination {@link QuadConstraintStream}'s tuple
+     * @return never null
+     */
+    <GroupKeyA_, GroupKeyB_, GroupKeyC_, GroupKeyD_>
+            QuadConstraintStream<GroupKeyA_, GroupKeyB_, GroupKeyC_, GroupKeyD_> groupBy(
+                    Function<A, GroupKeyA_> groupKeyAMapping, Function<A, GroupKeyB_> groupKeyBMapping,
+                    Function<A, GroupKeyC_> groupKeyCMapping, Function<A, GroupKeyD_> groupKeyDMapping);
+
+    // ************************************************************************
+    // Operations w/ duplicate tuple possibility
+    // ************************************************************************
+
+    /**
+     * Transforms the stream in such a way that tuples will be remapped using the given function.
+     * This may produce a stream with duplicate tuples.
+     * See {@link #distinct()} for details.
+     *
+     * There are several recommendations for implementing the mapping function:
+     *
+     * <ul>
+     * <li>Purity.
+     * The mapping function should only depend on its input.
+     * That is, given the same input, it always returns the same output.
+     * </li>
+     * <li>Simplicity.
+     * The function should only take the argument and return another simple value,
+     * without performing any significant computation.
+     * Not following this recommendation may result in performance loss,
+     * as the function may be called frequently.</li>
+     * <li>Bijectivity.
+     * No two input tuples should map to the same output tuple,
+     * or to tuples that are {@link Object#equals(Object) equal}.
+     * Not following this recommendation will create a constraint stream with duplicate tuples,
+     * and may force you to use {@link #distinct()} later.</li>
+     * <li>Immutable data carriers.
+     * The tuples returned by the mapping function should be identified by their contents and nothing else.
+     * They should also be immutable, meaning their contents should not be allowed to change.
+     * If two tuples carry objects which {@link Object#equals(Object) equal} one another,
+     * those two tuples should likewise {@link Object#equals(Object) equal} and preferably be the same instance.
+     * Java Records are the ideal tool for this particular use case.</li>
```

Comment:
Did we test Java Records for this? If it's not proven - we just think it is - we should probably leave it out.
I have done recommendations in the past that I hadn't tried out myself yet (although I was convinced they were a great idea) and people have treated those recommendations as something-that-should-be-questioned because the docs said it.